scipy/optimize/_numdiff.py

```diff
@@ -570,3 +638,48 @@ def check_derivative(fun, jac, x0, bounds=(-np.inf, np.inf), args=(),
                                    args=args, kwargs=kwargs)
         abs_err = np.abs(J_to_test - J_diff)
         return np.max(abs_err / np.maximum(1, np.abs(J_diff)))
+
+
+class FiniteDifference:
+    """Represents finite difference approximation.
+
+    Provide an easy way of calling `approx_derivative` repeatedly
+    with the same set of parameters. Avoid duplicated calls to ``fun``
+    if ``avoid_duplicated_calls=True``.
+    """
+
+    def __init__(self, fun, method='3-point', rel_step=None,
+                 bounds=(-np.inf, np.inf), sparsity=None,
+                 as_linear_operator=False, x0=None,
+                 avoid_duplicated_calls=True):
+        self._fun = fun
+        self.method = method
+        self.rel_step = rel_step
+        self.bounds = bounds
+        self.sparsity = sparsity
+        self.as_linear_operator = as_linear_operator
+        self.avoid_duplicated_calls = avoid_duplicated_calls
+        self._x = None
+        self._args = ()
+        self._kwargs = {}
+        self._f = None
+
+    def fun_wrapped(self, x, *args, **kwargs):
+        if not np.array_equal(self._x, x) or \
+           not args == self._args or \
```

Comment:
The problem is that this approach is not 100% robust, as you can't compare all possible object from args and kwargs by equality, on the other hand it will typically work as `args`, `kwargs` usually contain come scalar parameters for a problem. Perhaps explain this from documentation?

Another approach is to fix `args` and `kwargs` in the constructor, which I think is also practically sound, because one will rarely want to change them between calls. Actually I'm more fond of this approach.

The third approach is to disable this caching all together.

What do you think would be the best?